In [1]:
import pandas as pd
import numpy as np

In [2]:
data1 = pd.read_csv(r"C:\Users\94407\Downloads\customer_details.csv")
data2 =pd.read_csv(r"C:\Users\94407\Downloads\customer_policy_details.csv")


In [3]:
data1.columns= ["customer_id" ,
           "Gender","age",
           "driving licence present",
           "region code",
           "previously insured",
           "vehicle age",
           "vehicle damage"]
data2.columns = ["customer_id",
                 "annual premium (in Rs)",
                 "sales channel code",
                 "vintage" ,
                 "response" ]

### --------------------------------------------------------------------------------------------------------------------------------------------------------
### 2. Checking and Cleaning Data Quality:

i. Null values

Generate a summary of count of all the null values column wise

Drop Null values for customer_id because central tendencies for id’s is not feasible.

Replace all null values for numeric columns by mean. 

Replace all null values for Categorical value by mode.

In [4]:
df1 = pd.DataFrame(data = data1)
df2 = pd.DataFrame(data = data2)

df1.customer_id.isnull().sum()

386

In [5]:
df1.describe()
# Generate a summary of count of all the null values column wise

customer_id            age  driving licence present    region code  \
count  380723.000000  380741.000000            380716.000000  380717.000000   
mean   190548.776244      38.822827                 0.997867      26.388575   
std    110016.805160      15.512120                 0.046133      13.230361   
min         1.000000      20.000000                 0.000000       0.000000   
25%     95269.500000      25.000000                 1.000000      15.000000   
50%    190543.000000      36.000000                 1.000000      28.000000   
75%    285822.500000      49.000000                 1.000000      35.000000   
max    381109.000000      85.000000                 1.000000      52.000000   

       previously insured  
count       380728.000000  
mean             0.458246  
std              0.498254  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              1.000000  
max              1.000000

In [6]:
df2.describe()

customer_id  annual premium (in Rs)  sales channel code  \
count  380722.000000           380763.000000       380709.000000   
mean   190547.491663            30563.700273          112.035752   
std    110013.824148            17200.257877           54.204942   
min         1.000000             2630.000000            1.000000   
25%     95276.250000            24406.000000           29.000000   
50%    190536.500000            31668.000000          133.000000   
75%    285818.750000            39400.000000          152.000000   
max    381109.000000           540165.000000          163.000000   

             vintage       response  
count  380721.000000  380748.000000  
mean      154.342437       0.122543  
std        83.669600       0.327912  
min        10.000000       0.000000  
25%        82.000000       0.000000  
50%       154.000000       0.000000  
75%       227.000000       0.000000  
max       299.000000       1.000000

In [34]:
df1.head()

customer_id  Gender   age  driving licence present  region code  \
0          1.0    Male  44.0                      1.0         28.0   
1          2.0    Male  76.0                      1.0          3.0   
2          3.0    Male  47.0                      1.0         28.0   
3          4.0    Male  21.0                      1.0         11.0   
4          5.0  Female  29.0                      1.0         41.0   

   previously insured vehicle age vehicle damage  
0                 0.0   > 2 Years            Yes  
1                 0.0    1-2 Year             No  
2                 0.0   > 2 Years            Yes  
3                 1.0    < 1 Year             No  
4                 1.0    < 1 Year             No

In [36]:
print(df1.shape)
print(df2.shape)

(380723, 8)
(380722, 5)


In [7]:

# df1.describe()
# df1.customer_id.isnull().sum()


df1.dropna(subset=["customer_id"],inplace = True)

In [8]:
df1.isnull().sum()
# Drop Null values for customer_id because central tendencies for id’s is not feasible.

customer_id                  0
Gender                     368
age                        366
driving licence present    392
region code                391
previously insured         381
vehicle age                381
vehicle damage             406
dtype: int64

In [9]:
df1.age.fillna(df1.age.mean(),inplace = True)
df1["driving licence present"].fillna(df1["driving licence present"].mean(),inplace = True)
df1["region code"].fillna(df1["region code"].mean(),inplace = True)
df1["previously insured"].fillna(df1["previously insured"].mean(),inplace = True)
df1.isnull().sum()
# Replace all null values for numeric columns by mean.

customer_id                  0
Gender                     368
age                          0
driving licence present      0
region code                  0
previously insured           0
vehicle age                381
vehicle damage             406
dtype: int64

In [10]:
df1["Gender"].fillna(df1["Gender"].mode()[0],inplace = True)
df1["vehicle age"].fillna(df1["vehicle age"].mode()[0],inplace = True)
df1["vehicle damage"].fillna(df1["vehicle damage"].mode()[0],inplace = True)
df1.isnull().sum()

## customers_details table is cleaned (numeric values are filled with their column averages and cotegorial 
## columns null values are filled with their mode values)

customer_id                0
Gender                     0
age                        0
driving licence present    0
region code                0
previously insured         0
vehicle age                0
vehicle damage             0
dtype: int64

In [11]:
#  Lets do the same with the customer policy details table also
print("The null values in customer policy table : \n",df2.isnull().sum())
# Drop Null values for customer_id because central tendencies for id’s is not feasible.
df2.dropna(subset=["customer_id"],inplace = True)
print("The null values in customer policy table : \n",df2.isnull().sum())

The null values in customer policy table : 
 customer_id               387
annual premium (in Rs)    346
sales channel code        400
vintage                   388
response                  361
dtype: int64
The null values in customer policy table : 
 customer_id                 0
annual premium (in Rs)    344
sales channel code        400
vintage                   388
response                  361
dtype: int64


In [12]:
# Replace all null values for numeric columns by mean.

df2["annual premium (in Rs)"].fillna(df2["annual premium (in Rs)"].mean(),inplace = True)

In [13]:
df2.head()
print("The null values in customer policy table : \n",df2.isnull().sum())

The null values in customer policy table : 
 customer_id                 0
annual premium (in Rs)      0
sales channel code        400
vintage                   388
response                  361
dtype: int64


In [14]:
df2["sales channel code"].value_counts().count() # sales channel code column has 155 values == > categorial
df2["vintage"].value_counts().count() # vintage column has 290 values == > categorial
df2["response"].value_counts().count() # vintage column has 2 values == > categorial
#  So we replace the null values with the mode of the column
print("The null values in customer policy table : \n",df2.isnull().sum())

The null values in customer policy table : 
 customer_id                 0
annual premium (in Rs)      0
sales channel code        400
vintage                   388
response                  361
dtype: int64


In [15]:
df2["sales channel code"].fillna(df2["sales channel code"].mode()[0],inplace = True)
df2["vintage"].fillna(df2["vintage"].mode()[0],inplace = True)
df2["response"].fillna(df2["response"].mode()[0],inplace = True)
print("The null values in customer policy table : \n",df2.isnull().sum())

The null values in customer policy table : 
 customer_id               0
annual premium (in Rs)    0
sales channel code        0
vintage                   0
response                  0
dtype: int64


Cleaning the data from both the tables has finished

ii. Outliers

Generate a summary of count of all the outliers column wise
Replace all outlier values for numeric columns by mean. 
(Hint1: for outlier treatment use IQR method as follows:

For example: for a column X calculate Q1 = 25th percentile and Q3 = 75th percentile then IQR = Q3 – Q1 ) then to check outlier, anything lower than a Q1 – 1.5IQR or greater than Q3 + 1.5 IQR would be an outlier

Hint2: For getting percentile value, explore pd.describe() function)

In [16]:
df1.describe()

customer_id            age  driving licence present    region code  \
count  380723.000000  380723.000000            380723.000000  380723.000000   
mean   190548.776244      38.822788                 0.997868      26.389436   
std    110016.805160      15.504826                 0.046104      13.223671   
min         1.000000      20.000000                 0.000000       0.000000   
25%     95269.500000      25.000000                 1.000000      15.000000   
50%    190543.000000      36.000000                 1.000000      28.000000   
75%    285822.500000      49.000000                 1.000000      35.000000   
max    381109.000000      85.000000                 1.000000      52.000000   

       previously insured  
count       380723.000000  
mean             0.458259  
std              0.498006  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              1.000000  
max              1.000000

In [17]:
# q1_age,q3_age = np.percentile(df1.age,[25,75])
# print(q1_age,q3_age)
# q1_dlp ,q3_dlp = np.percentile(df1["driving licence present"],[25,75])
# # print(q1_dlp,q2_dlp)

In [18]:
# iqr = q3_age -q1_age
# lower_iqr_range = q1_age -(1.5 * iqr)
# higher_iqr_range = q3_age +(1.5 * iqr)
# print(lower_iqr_range ,higher_iqr_range)# There are no -ve age in the age column

In [19]:
# df1[df1.age > higher_iqr_range].age.sum() # there are 0 outliers in the age column

In [20]:
def iqr_range(col):
    q1,q3 = np.percentile(col,[25,75]) 
#     print(q1,q3)
    iqr = q3-q1
    low_iqr_lvl = q1 -(1.5*iqr)
    high_iqr_lvl = q3 + (1.5*iqr)   
#     print("Higher IQR Range : " ,high_iqr_lvl )
#     print("Lower IQR Range : " ,low_iqr_lvl )
    return [low_iqr_lvl,high_iqr_lvl]
    
    

iqr_range(df1.age)
print("No. Outliers greater then in age column :\t",df1[df1["age"] > iqr_range(df1.age)[1] ].age.sum())
print("No. Outliers lesser then in age column :\t",df1[df1["age"] < iqr_range(df1.age)[0] ].age.sum())
# There are 0 outliers in Age column

No. Outliers greater then in age column :	 0.0
No. Outliers lesser then in age column :	 0.0


In [21]:
print(df1["vehicle age"].value_counts())
print(df1["driving licence present"].value_counts())
# I didnot implement iqr on other columns because allother columns are categorial type 

1-2 Year     200283
< 1 Year     164468
> 2 Years     15972
Name: vehicle age, dtype: int64
1.000000    379520
0.000000       811
0.997868       392
Name: driving licence present, dtype: int64


lets find the number of outliers in the policy details table

In [22]:
df2.describe()

customer_id  annual premium (in Rs)  sales channel code  \
count  380722.000000           380722.000000       380722.000000   
mean   190547.491663            30563.999774          112.078674   
std    110013.824148            17190.147550           54.192513   
min         1.000000             2630.000000            1.000000   
25%     95276.250000            24416.000000           29.000000   
50%    190536.500000            31656.000000          136.000000   
75%    285818.750000            39391.750000          152.000000   
max    381109.000000           540165.000000          163.000000   

             vintage       response  
count  380722.000000  380722.000000  
mean      154.450788       0.122410  
std        83.690971       0.327759  
min        10.000000       0.000000  
25%        82.000000       0.000000  
50%       154.000000       0.000000  
75%       227.000000       0.000000  
max       299.000000       1.000000

In [23]:
iqr_range(df2["annual premium (in Rs)"])

[1952.375, 61855.375]

In [24]:
print("No. Outliers greater then in annual premium (in Rs) column :\t",df2[df2["annual premium (in Rs)"] > iqr_range(df2["annual premium (in Rs)"])[1] ]["annual premium (in Rs)"].count())
print("No. Outliers lesser  then in annual premium (in Rs) column :\t",df2[df2["annual premium (in Rs)"] < iqr_range(df2["annual premium (in Rs)"])[0] ]["annual premium (in Rs)"].count())
# There are 0 outliers in Age column

No. Outliers greater then in annual premium (in Rs) column :	 10332
No. Outliers lesser  then in annual premium (in Rs) column :	 0


### There are zero outliers in the customer detail (age column)
### but we get 10332 values in the policy details (annual premium (in Rs) column) 

### --------------------------------------------------------------------------------------------------------------------------------------------------------
### 3. Create a Master table for future use. Join the customer table and customer_policy table to get a master table using customer_id in both tables.

(Hint: use pd.merge() function)

In [25]:
df1.head()

customer_id  Gender   age  driving licence present  region code  \
0          1.0    Male  44.0                      1.0         28.0   
1          2.0    Male  76.0                      1.0          3.0   
2          3.0    Male  47.0                      1.0         28.0   
3          4.0    Male  21.0                      1.0         11.0   
4          5.0  Female  29.0                      1.0         41.0   

   previously insured vehicle age vehicle damage  
0                 0.0   > 2 Years            Yes  
1                 0.0    1-2 Year             No  
2                 0.0   > 2 Years            Yes  
3                 1.0    < 1 Year             No  
4                 1.0    < 1 Year             No

In [26]:
df2.head()

customer_id  annual premium (in Rs)  sales channel code  vintage  response
0          1.0                 40454.0                26.0    217.0       1.0
1          2.0                 33536.0                26.0    183.0       0.0
2          3.0                 38294.0                26.0     27.0       1.0
3          4.0                 28619.0               152.0    203.0       0.0
4          5.0                 27496.0               152.0     39.0       0.0

In [27]:
m_table = pd.merge(df1,df2,on= "customer_id")
m_table.head()

customer_id  Gender   age  driving licence present  region code  \
0          1.0    Male  44.0                      1.0         28.0   
1          2.0    Male  76.0                      1.0          3.0   
2          3.0    Male  47.0                      1.0         28.0   
3          4.0    Male  21.0                      1.0         11.0   
4          5.0  Female  29.0                      1.0         41.0   

   previously insured vehicle age vehicle damage  annual premium (in Rs)  \
0                 0.0   > 2 Years            Yes                 40454.0   
1                 0.0    1-2 Year             No                 33536.0   
2                 0.0   > 2 Years            Yes                 38294.0   
3                 1.0    < 1 Year             No                 28619.0   
4                 1.0    < 1 Year             No                 27496.0   

   sales channel code  vintage  response  
0                26.0    217.0       1.0  
1                26.0    183.0       0.0  
2                26.0     27.0       1.0  
3               152.0    203.0       0.0  
4               152.0     39.0       0.0

### --------------------------------------------------------------------------------------------------------------------------------------------------------
### 4. Company needs some important information from the master table to make decisions for future growth.They needs following information:

 i. Gender wise average annual premium

ii. Age wise average annual premium

iii. Is your data balanced between the genders?

          (Hint: Data is balanced if number of counts in each group is approximately same)

iv. Vehicle age wise average annual premium.

In [28]:
# i. Gender wise average annual premium
m_table.groupby(by = ["Gender"])["annual premium (in Rs)"].mean()

Gender
Female    30492.028478
Male      30623.619273
Name: annual premium (in Rs), dtype: float64

In [29]:
# ii. Age wise average annual premium
m_table.groupby(by = ["age"])["annual premium (in Rs)"].mean()

age
20.0    26924.620173
21.0    30564.475810
22.0    30823.778102
23.0    30688.606298
24.0    31183.802890
            ...     
81.0    31201.571429
82.0    37705.379310
83.0    31012.727273
84.0    35440.818182
85.0    29792.363636
Name: annual premium (in Rs), Length: 67, dtype: float64

In [30]:
# iii. Is your data balanced between the genders?
m_table.groupby(by = ["Gender"]).count()["customer_id"]

#  No The data is not balanced between the genders

Gender
Female    174485
Male      205851
Name: customer_id, dtype: int64

In [31]:
m_table.head()

customer_id  Gender   age  driving licence present  region code  \
0          1.0    Male  44.0                      1.0         28.0   
1          2.0    Male  76.0                      1.0          3.0   
2          3.0    Male  47.0                      1.0         28.0   
3          4.0    Male  21.0                      1.0         11.0   
4          5.0  Female  29.0                      1.0         41.0   

   previously insured vehicle age vehicle damage  annual premium (in Rs)  \
0                 0.0   > 2 Years            Yes                 40454.0   
1                 0.0    1-2 Year             No                 33536.0   
2                 0.0   > 2 Years            Yes                 38294.0   
3                 1.0    < 1 Year             No                 28619.0   
4                 1.0    < 1 Year             No                 27496.0   

   sales channel code  vintage  response  
0                26.0    217.0       1.0  
1                26.0    183.0       0.0  
2                26.0     27.0       1.0  
3               152.0    203.0       0.0  
4               152.0     39.0       0.0

In [32]:
# iv. Vehicle age wise average annual premium.
m_table.groupby(by = "vehicle age")["annual premium (in Rs)"].mean()

vehicle age
1-2 Year     30524.629840
< 1 Year     30115.716367
> 2 Years    35657.520845
Name: annual premium (in Rs), dtype: float64

### --------------------------------------------------------------------------------------------------------------------------------------------------------
### 5. Is there any relation between Person Age and annual premium?

Hint: use correlation function (Correlation describes the relationship between two variables). 

Correlation coefficient < -0.5           - Strong negative relationship

Correlation coefficient > 0.5            -  Strong positive relationship

0.5 < Correlation coefficient < 0.5   - There is no relationship. 

In [33]:
corr = m_table["age"].corr(m_table["annual premium (in Rs)"])
print(corr)

0.06771515986613927


Correlation coefficient  = 0.067715

#### correlation coefficient is between  -0.5 and 0.5 ==> No Relationship

Which is less than 0.5